xxxx-gpu-msm.md

```diff
@@ -0,0 +1,92 @@
+# RFC title
+
+GPU based MSM implementation.
+
+## Summary
+
+A rust library implementing gpu based acceleration for computing MSM, it can be integrated in kimchi and
+other projects that make use of MSMs with the pasta curves.
+
+## Motivation
+
+One of the mains components of kimchi's runtime is the multi scalar multiplication (MSM), it consist of
+basically taking a list of points, adding each point to itself N times, and then add all the resulting
+points together.
+This library will accelerate that process doing the computation in a gpu, it will also serve as a reference
+point to measure the improvements that gpu computations can provide us.
+
+This will build on top of an already existing implementation, and thus some of the design choices will be
+based on the fact that the particular decision is cheap to implement because it is already implemented.
+
+## Detailed design
+
+The implementation will be based in the wgpu graphics API and thus the WGSL shader language. The library
+will consist of 3 crates:
+    - arithmetic
+    - shader
+    - runner
+
+### arithmetic
+
+This is a crate implementing general field arithmetic, general curve arithmetic and specifically the
+pasta curves and their field. It uses no dependencies and avoids hardware features that are unsupported
+by GPUs in general and WGSL in particular.
+The code is also written in a way optimized for GPUs, but it can still be used in any architecture, while
+it will be generally slower that other cpu implementations it can be used for things like tests.
+
+### shader
+
+The shader is a small crate that will be compiled to WGSL, the code is very simple, mostly calls to
+functions from the arithmetic crate, compilation takes a significative time and thus the compiled
+shader will be saved and the runner will used that precompiled shader by default.
+
+### runner
+
+This is library that will be used by end users of the crate, it is the part that will run in the CPU
+and will talk to the GPU, it will expose a simple API that will ask for the points that make the base
+of the MSM and then perform the initialization and provide some struct that can be used to send scalar
+and get back the MSM of them and the base sent in the initialization.
+
+## Test plan and functional requirements
+
+Given that the goal is more to see the kind of performance gains we can obtain, the testing can be
+limited to asserting that the result obtained is the same as our current CPU implementation.
+Before this becomes the default implementation to be used in production we will require a considerable
+amount of reviewing and additional test considering that the arithmetic crate is an implementation
+from scratch of the basic cryptographic building blocks.
+There are already some unit tests and some tests using formal verification, but the formal verification
+tests take a while to run a probably can't cover all cases.
+
+## Drawbacks
```

Comment:
The details in this section feel like they belong more in the 'functional requirements' above. E.g. `compatibility with nvidia Vulkan is out-of-scope for this prototype, due to driver issues`

Reply:
It was just an example, the reason Vulkan is not a target is just because it won't work in browser, we can maybe have a non-browser specialized version that uses Vulkan after this prototype.
Do you still want some change to either section?

Reply:
It's just a nit really, it's probably fine to leave as-is